ssl.c

```diff
@@ -92,142 +92,172 @@ void
 		  uint8 * exponent)
 {
-	BN_CTX *ctx;
-	BIGNUM *mod, *exp, *x, *y;
-	uint8 inr[SEC_MAX_MODULUS_SIZE];
-	int outlen;
-
-	reverse(modulus, modulus_size);
-	reverse(exponent, SEC_EXPONENT_SIZE);
-	memcpy(inr, in, len);
-	reverse(inr, len);
-
-	ctx = BN_CTX_new();
-	mod = BN_new();
-	exp = BN_new();
-	x = BN_new();
-	y = BN_new();
-
-	BN_bin2bn(modulus, modulus_size, mod);
-	BN_bin2bn(exponent, SEC_EXPONENT_SIZE, exp);
-	BN_bin2bn(inr, len, x);
-	BN_mod_exp(y, x, exp, mod, ctx);
-	outlen = BN_bn2bin(y, out);
-	reverse(out, outlen);
+	mpz_t exp, mod;
+
+	mpz_t y;
+	mpz_t x;
+
+	size_t outlen;
+
+	mpz_init(y);
+	mpz_init(x);
+	mpz_init(exp);
+	mpz_init(mod);
+
+	mpz_import(mod, modulus_size, 1, sizeof(modulus[0]), 0, 0, modulus);
+	// TODO: Need exponent size
+	mpz_import(exp, 3, 1, sizeof(exponent[0]), 0, 0, exponent);
+
+	mpz_import(x, len, -1, sizeof(in[0]), 0, 0, in);
+
+	mpz_powm(y, x, exp, mod);
+
+	mpz_export(out, &outlen, -1, sizeof(out[0]), 0, 0, y);
+
 	if (outlen < (int) modulus_size)
 		memset(out + outlen, 0, modulus_size - outlen);
-
-	BN_free(y);
-	BN_clear_free(x);
-	BN_free(exp);
-	BN_free(mod);
-	BN_CTX_free(ctx);
 }
 
 /* returns newly allocated RDSSL_CERT or NULL */
 RDSSL_CERT *
 rdssl_cert_read(uint8 * data, uint32 len)
 {
-	/* this will move the data pointer but we don't care, we don't use it again */
-	return d2i_X509(NULL, (D2I_X509_CONST unsigned char **) &data, len);
+	int ret;
+	gnutls_datum_t cert_data;
+	gnutls_x509_crt_t *cert;
+
+	cert = malloc(sizeof(*cert));
+
+	if (!cert) {
+		logger(Protocol, Error, "%s:%s:%d: Failed to allocate memory for certificate structure.\n",
+				__FILE__, __func__, __LINE__);
+		return NULL;
+	}
+
+	if ((ret = gnutls_x509_crt_init(cert)) != GNUTLS_E_SUCCESS) {
+		logger(Protocol, Error, "%s:%s:%d: Failed to init certificate structure. GnuTLS error = 0x%02x (%s)\n",
+				__FILE__, __func__, __LINE__, ret, gnutls_strerror(ret));
+
+		return NULL;
+	}
+
+	cert_data.size = len;
+	cert_data.data = data;
+
+	if ((ret = gnutls_x509_crt_import(*cert, &cert_data, GNUTLS_X509_FMT_DER)) != GNUTLS_E_SUCCESS) {
+		logger(Protocol, Error, "%s:%s:%d: Failed to import DER encoded certificate. GnuTLS error = 0x%02x (%s)\n",
+				__FILE__, __func__, __LINE__, ret, gnutls_strerror(ret));
+		return NULL;
+	}
+
+	return cert;
 }
 
 void
 rdssl_cert_free(RDSSL_CERT * cert)
 {
-	X509_free(cert);
+	gnutls_free(cert);
 }
 
+
+/*
+ * AFAIK, there's no way to alter the decoded certificate using GnuTLS.
+ *
+ * Upon detecting "problem" (wrong public RSA key OID) certificate
+ * we basically have two options:
+ *
+ * 1)) encode certificate back to DER, then parse it using libtasn1,
+ * fix public key OID (set it to 1.2.840.113549.1.1.1), encode to DER again
+ * and finally reparse using GnuTLS
+ *
+ * 2) encode cert back to DER, get RSA public key parameters using libtasn1
+ *
+ * Or can rewrite the whole certificate related stuff later.
```

Comment:
Any tweak flags or something to GnuTLS that can make it compensate for this? Their developer is usually very responsive so it could be worth asking on their mailing list.

Reply:
Alright, let's go with this for now